### DISCUSSION TITLE
[Freature Request] Event for idling (_km_idle_timeout)

### DISCUSSION THREAD

**Discussion by j16sdiz** [jschuh/klipper-macros#244] [Ideas]
Expose the Timeout Idle for event.

Reasonale: I have NEOPIXEL LEDs, would be great if they can dim after idle.

**Comment by gretel** [comment]
think that is possible rn? https://github.com/jschuh/klipper-macros/blob/14840a4f1cdc38358d3c2097088e58ac5d502914/README.md?plain=1#L220

**Comment by j16sdiz** [comment]
@gretel I expect that's something like `GCODE_ON_PRINT_STATUS` that can be set runtime (i.e. without restarting klippy)

**Comment by jschuh** [comment]
The problem is that Klipper currently doesn't provide a good way for a macro to detect that the printer has transitioned out of the idle state. So, in the case of dimming LEDs when the printer is idle, you would also expect those LEDs to return to their previous state as soon as you issue _any_ command. Unfortunately there's no good way to do that, so the best you can do is kinda fake it with ugly hacks.

The easiest hack is to just ignore the problem, and hope an overridden command eventually gets issued and sets the state after the printer comes out of idle. The problem with that approach is that it's going to be inconsistent, since only a handful of commands are hooked to track the state. So you don't actually know if or when your LEDs will come back on.

The more complicated hack is to use `delayed_gcode` to poll the `idle_timeout.state` every second or so while the printer is in the idle state. The first problem is that sometimes an overridden command will bring you out of the idle state before the `delayed_gcode` fires, so you'll need some fiddly logic to handle that corner case. The other problem is that you'll be constantly running gcode while the printer is supposed to be idle, which just seems like a bad idea since that's when the host really should be minimizing its workload and ideally going into a low power mode.

Since I didn't like either of those approaches, I chose not to support an idle state.
